### PR TITLE
fees: clearer error message

### PR DESCRIPTION
**Pull request by TalDerei**
references https://github.com/penumbra-zone/web/issues/1413 and https://github.com/prax-wallet/web/issues/148

Screenshot 2024-08-18 at 11 37 45 PM

### CODE REVIEW

**Comment by grod220** [comment]
This also seems to address: https://github.com/prax-wallet/web/issues/148

File: packages/wasm/crate/src/planner.rs [hunk 1]
```diff
@@ -601,13 +601,7 @@ pub async fn plan_transaction_inner<Db: Database>(
...
             .get_mut(&required.asset_id)
             .and_then(|notes| notes.pop())
-            .ok_or_else(|| {
-                anyhow!(
-                    "Failed to retrieve or ran out of notes for asset {}, required amount {}",
-                    required.asset_id,
-                    required.amount
-                )
-            })?;
+            .ok_or_else(|| anyhow!("Transaction failed due to insufficient funds"))?;
```
[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/wasm/crate/src/planner.rs:604]: I think it’s a good idea, but how do we display UM? For instance, logging these variables returns

```
let asset: Metadata = storage.get_asset(&required.asset_id).await?.unwrap();
let asset_id = asset.id();
let asset_denom = asset.base_denom();
let asset_unit = asset.base_unit();
```

Screenshot 2024-08-19 at 8 08 57 PM

I don't see a getter for retrieving the [name](https://github.dev/penumbra-zone/penumbra/blob/56cb0025e0cece24a6301d72d46bfe7094266daf/crates/core/asset/src/asset/denom_metadata.rs#L49-L50) field?

[the remark this answers is not in the document]
Comment (grod220) [reply packages/wasm/crate/src/planner.rs:604]: What do you think about this?

```rust
    while let Some(required) = actions_list.balance_with_fee().required().next() {
        // Find a single note to spend towards the required balance.
        let maybe_note = notes_by_asset_id
            .get_mut(&required.asset_id)
            .and_then(|notes| notes.pop());
        let note = match maybe_note {
            Some(note) => Ok(note),
            None => Err(insufficient_funds_err(&storage, &required).await),
        }?;
       // .....
    }
```
And then a helper function:
```rust
// Used to augment error message with denom metadata for a more readable error
async fn insufficient_funds_err<Db: Database>(
    storage: &Storage<Db>,
    required: &Value,
) -> WasmError {
    let metadata_result = storage.get_asset(&required.asset_id).await;
    let error_message = match metadata_result {
        Ok(Some(asset_metadata)) => {
            let display_unit = asset_metadata.default_unit();
            format!(
                "Transaction failed due to insufficient funds. Required: {}{}",
                display_unit.format_value(required.amount),
                display_unit
            )
        }
        // Gracefully handle cases when database errors or does not have metadata for asset
        _ => "Transaction failed due to insufficient funds".to_string(),
    };
    anyhow!(error_message).into()
}

```

**Comment by grod220** [comment]
Also, we should have tests for this please 🙏

File: packages/wasm/crate/src/planner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/wasm/crate/src/planner.rs:604]: this is great! I was looking for `display_unit`

[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/wasm/crate/src/planner.rs:604]: Screenshot 2024-08-19 at 8 08 41 PM

[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/wasm/crate/src/planner.rs:604]: Screenshot 2024-08-20 at 12 13 47 PM

[the remark this answers is not in the document]
Comment (TalDerei) [reply packages/wasm/crate/src/planner.rs]: need to remove transfer channels for IBC assets 

Screenshot 2024-08-20 at 12 39 56 PM

File: packages/wasm/crate/src/planner.rs [hunk 2]
```diff
@@ -137,6 +137,28 @@ async fn save_metadata_if_needed<Db: Database>(
...
+    storage: &Storage<Db>,
+    required: &Value,
+) -> WasmError {
+    let metadata_result = storage.get_asset(&required.asset_id).await;
+    let error_message = match metadata_result {
+        Ok(Some(asset_metadata)) => {
+            let display_unit = asset_metadata.default_unit();
+            format!(
+                "Transaction failed due to insufficient funds. 
+                Required amount: {} {}",
```
Comment (grod220) [inline packages/wasm/crate/src/planner.rs:151]: nit: can you add a period?

```diff
- Required amount: {} {}
+ Required amount: {} {}.
```

File: packages/wasm/crate/tests/test_insufficient_notes.rs
Comment (grod220) [inline packages/wasm/crate/tests/test_insufficient_notes.rs:1]: issue: we need at least two tests (maybe three):

- Asset id metadata is not in storage (show backup generic error)
- Asset id metadata is in storage (show the display denom + amount) 
- Db throws an error while retrieving (show backup generic error). This last one may be tricky to pull off with the mock database. Feel free to exclude it too difficult.

File: packages/wasm/crate/src/planner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (grod220) [reply packages/wasm/crate/src/planner.rs]: We can leave the ibc denom's as is. If it proves confusing to users, can revisit. But regardless, far more clear than the current error.

File: packages/wasm/crate/tests/test_insufficient_notes.rs
Comment (TalDerei) [reply packages/wasm/crate/tests/test_insufficient_notes.rs:1]: expanded testing coverage

File: packages/wasm/crate/tests/test_insufficient_notes.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    let key = byte_array_to_base64(&asset_id_staking_token.to_proto().inner);
+    let metadata_in_db_proto = pb::Metadata {
+        base: asset_id_staking_token.to_string(),
+        ..Default::default()
+    };
+
+    mock_db
+        .put_with_key(&tables.assets, key.clone(), &metadata_in_db_proto)
+        .await
+        .unwrap();
```
Comment (grod220) [inline packages/wasm/crate/tests/test_insufficient_notes.rs:109]: question: this is not necessary for this test right?

File: packages/wasm/crate/tests/test_insufficient_notes.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    let tables = get_mock_tables();
+    seed_params_in_db(&mock_db, &tables).await;
+
+    let asset_id_staking_token = *STAKING_TOKEN_ASSET_ID;
+
+    let key = byte_array_to_base64(&asset_id_staking_token.to_proto().inner);
+    let metadata_in_db_proto = pb::Metadata {
+        base: asset_id_staking_token.to_string(),
+        ..Default::default()
+    };
```
Comment (grod220) [inline packages/wasm/crate/tests/test_insufficient_notes.rs:65]: addon: interestingly, you found another case I didn't think about. That is: what if the database has metadata but no display info? Think it's worth having an additional test for this.

Revised test case:
```rust
async fn metadata_in_storage_but_no_display_info() {
...
// The current code you are using
    let metadata_in_db_proto = pb::Metadata {
        base: asset_id_staking_token.to_string(),
        ..Default::default()
    };
...
 assert_eq!(
        format!("{}", error),
       // I'd consider putting the whole id there so it's clear to folks
         "Transaction failed due to insufficient funds. Required amount: 100 passet1984fctenw8m2fpl8a9wzguzp7j34d7vravryuhft808nyt9fdggqxmanqm."
    );

```

Additional one:
```rust
async fn metadata_in_storage() {
...
// Store whole metadata struct
let metadata_in_db_proto: pb::Metadata = STAKING_TOKEN_DENOM.clone().into();
..
    assert_eq!(
        format!("{}", error),
       // note the decimal change
        "Transaction failed due to insufficient funds. Required amount: 0.0001 penumbra."
    );
```

File: packages/wasm/crate/tests/test_insufficient_notes.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+use penumbra_wasm::database::mock::{get_mock_tables, MockDb};
+use penumbra_wasm::planner::insufficient_funds_err;
+use penumbra_wasm::storage::{byte_array_to_base64, Storage};
+use rand_core::OsRng;
+use wasm_bindgen_test::wasm_bindgen_test;
+wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+use penumbra_proto::DomainType;
+
+#[wasm_bindgen_test]
+// Database throws an error while retrieving
```
Comment (grod220) [inline packages/wasm/crate/tests/test_insufficient_notes.rs:18]: issue: this test hits the same case as `test_insufficient_notes_not_in_storage`. Triggering a database error is actually not supported in our mock db right now. To not confuse others, I'd leave a comment like:

```
// TODO: add a test for database error when mock storage can support errors
```

**Review by grod220 — APPROVED** [review]
Approving, but left some comments for another iteration on some things in the tests

File: packages/wasm/crate/tests/test_insufficient_notes.rs (hunk 3 above)
Comment (TalDerei) [reply packages/wasm/crate/tests/test_insufficient_notes.rs:109]: good catch, this was a copy of the other test.

File: packages/wasm/crate/tests/test_insufficient_notes.rs (hunk 4 above)
Comment (TalDerei) [reply packages/wasm/crate/tests/test_insufficient_notes.rs:65]: added the additional test case

**Pull request merged**